drivers/adc/adc_stm32.c

```diff
@@ -316,14 +316,18 @@ static int adc_stm32_dma_start(const struct device *dev,
 	blk_cfg = &dma->dma_blk_cfg;
 
 	/* prepare the block */
-	blk_cfg->block_size = channel_count * sizeof(int16_t);
+	blk_cfg->block_size = data->channel_count * sizeof(int16_t);
+	if (ctx->options.batch_mode) {
```

Comment:
I'm afraid we need to limit the batch mode to those STM32 series that can stop continuous mode conversion without disabling the whole peripheral. That is, unless we could find a way to disable and reenable it quickly inside interrupt handler.